src/mpyl/cli/projects.py

```diff
@@ -159,5 +167,48 @@ def lint(obj: ProjectsContext, all_, extended):
         )
 
 
+@projects.command(help="Upgrade projects to conform with the latest schema")
+@click.option(
+    "--apply",
+    "-a",
+    is_flag=True,
+    help="Apply upgrade operations to the project files",
+)
+@click.pass_obj
+def upgrade(obj: ProjectsContext, apply):
+    paths = map(Path, _find_project_paths(True, obj.cli.repo, ""))
+    candidates = check_upgrades_needed(list(paths))
+    if not apply:
+        upgradable = check_upgrade(obj.cli.console, candidates)
+        number_in_need_of_upgrade = len(upgradable)
+        if number_in_need_of_upgrade > 0:
+            obj.cli.console.print(
+                f"{number_in_need_of_upgrade} projects need to be upgraded"
+            )
+            sys.exit(1)
+
+    with obj.cli.console.status("Checking for upgrades...") as status:
+        materialized = list(candidates)
+        need_upgrade = [path for path, diff in materialized if diff is not None]
+        number_of_upgrades = len(need_upgrade)
```

Comment:
This logic can be merged with the logic on lines 182-183 I would think? Not sure (yet) why they use different ways to get to the same result.

Reply:
It's slightly different. When we do the check, we log while processing the items in the generator. This gives us a streaming output https://github.com/Vandebron/mpyl/blob/6617b612f85a50d437db18b162f9247eae46cbd5/src/mpyl/cli/commands/projects/upgrade.py#L14

When applying the upgrades, we need to materialize the generator first. Because we don't need to report progress and just want to show how many candidates there are.